refactor(*): backport `coe_fn` from Lean 4

urkud opened the pull request:
See leanprover-community/lean#557 for Lean core part. There are a few remaining bugs:

- changes in the elaboration order made some code fail;
- `@[derive has_coe_to_fun]` is no longer possible;
- Lean fails to use `coe_fn_coe_base` as a `simp` lemma;
- some tests are not fixed yet.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

urkud committed:
Snapshot

urkud committed:
Switch to a new tag

urkud committed:
Merge branch 'master' into coe-fn-backport

urkud committed:
migrate more

urkud committed:
Fix simps

urkud committed:
Fix here and there

urkud committed:
Fix

urkud committed:
Snapshot

urkud committed:
Merge branch 'master' into coe-fn-backport

urkud committed:
...

b-mehta committed:
try fix elementwise

b-mehta committed:
fix more?

b-mehta committed:
fix more

urkud committed:
Fix more

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Fix some tests

urkud committed:
Update lean

urkud committed:
Merge branch 'master' into coe-fn-backport

# Conflicts:
#	leanpkg.toml
#	src/algebra/category/Group/basic.lean
#	src/algebra/category/Mon/basic.lean
#	src/algebra/category/Semigroup/basic.lean
#	src/algebra/group/hom.lean
#	src/algebra/group_action_hom.lean
#	src/analysis/analytic/basic.lean
#	src/analysis/normed_space/dual.lean
#	src/analysis/normed_space/enorm.lean
#	src/analysis/normed_space/linear_isometry.lean
#	src/computability/partrec_code.lean
#	src/data/equiv/basic.lean
#	src/data/equiv/mul_add.lean
#	src/data/mv_polynomial/equiv.lean
#	src/data/nat/prime.lean
#	src/data/polynomial/basic.lean
#	src/field_theory/polynomial_galois_group.lean
#	src/geometry/manifold/bump_function.lean
#	src/group_theory/order_of_element.lean
#	src/linear_algebra/affine_space/affine_equiv.lean
#	src/linear_algebra/alternating.lean
#	src/linear_algebra/quadratic_form.lean
#	src/linear_algebra/special_linear_group.lean
#	src/logic/embedding.lean
#	src/measure_theory/integral/bochner.lean
#	src/measure_theory/measure/measure_space.lean
#	src/order/closure.lean
#	src/ring_theory/derivation.lean
#	src/ring_theory/fractional_ideal.lean
#	src/set_theory/pgame.lean
#	src/set_theory/zfc.lean
#	src/tactic/elementwise.lean
#	src/tactic/simps.lean
#	src/topology/category/CompHaus.lean
#	src/topology/category/Profinite.lean
#	src/topology/category/Top/open_nhds.lean
#	src/topology/category/Top/opens.lean
#	src/topology/metric_space/gromov_hausdorff.lean
#	src/topology/path_connected.lean
#	src/topology/topological_fiber_bundle.lean
#	src/topology/vector_bundle.lean
#	test/simps.lean

urkud committed:
Snapshot

urkud committed:
Merge branch 'master' into coe-fn-backport

urkud committed:
Update lean

urkud committed:
Revert whiespace

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -466,11 +466,13 @@ instance coe_add_monoid_hom : has_coe (A →ₐ[R] B) (A →+ B) := ⟨λ f, ↑
-theorem coe_fn_inj ⦃φ₁ φ₂ : A →ₐ[R] B⦄ (H : ⇑φ₁ = φ₂) : φ₁ = φ₂ :=
-by { cases φ₁, cases φ₂, congr, exact H }
+theorem coe_fn_injective : function.injective (coe_fn : (A →ₐ[R] B) → (A → B)) :=
+by { intros φ₁ φ₂ H, cases φ₁, cases φ₂, congr, exact H }
```
I've extracted this lemma to #9241

eric-wieser commented on src/algebra/category/CommRing/basic.lean:
```diff
@@ -67,7 +69,9 @@ namespace Ring
-attribute [derive [has_coe_to_sort, large_category, concrete_category]] Ring
+attribute [derive [large_category, concrete_category]] Ring
+
+instance : has_coe_to_sort Ring Type* := bundled.has_coe_to_sort
```
Does the following work?
```suggestion
attribute [derive [(λ Ring, has_coe_to_sort Ring Type*), large_category, concrete_category]] Ring
```

Vierkantor committed:
Update `combinatorics/quiver.lean`

Vierkantor committed:
Fix `data/finset/basic.lean`

Vierkantor committed:
Fix `data/fintype/basic.lean`

Vierkantor committed:
Use a `derive` to get `coe_to_sort`

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

urkud committed:
Fix some compile failures

urkud committed:
Merge branch 'coe-fn-backport' of git://github.com/leanprover-community/mathlib into coe-fn-backport

Vierkantor committed:
Fix `number_theory/dioph.lean`

urkud committed:
Fix data.real.basic

Vierkantor committed:
Fix `group_theory/perm/cycles.lean`

Vierkantor committed:
Fix `combinatorics/derangements/basic.lean`

Vierkantor commented on src/group_theory/perm/cycles.lean:
```diff
@@ -129,7 +129,8 @@ noncomputable def is_cycle.gpowers_equiv_support {σ : perm α} (hσ : is_cycle
 begin
   obtain ⟨τ, n, rfl⟩ := τ,
-  rw [finset.mem_coe, coe_fn_coe_base, subtype.coe_mk, gpow_apply_mem_support, mem_support],
+  rw [finset.mem_coe, @coe_fn_coe_base _ _ (λ _, α → α), subtype.coe_mk,
```
It seems `rw` and `simp` are not smart enough to figure out the `(λ _, α → α)` (even if it's an `out_param`). Should we add a non-dependent `coe_fn_coe_base'` where the argument would just be `α → α`?

urkud replied:
May be `α → β` will work as well. I don't know why `coe_fn_coe_base'` doesn't work. I'll be very glad if you'll fix it one way or another.

Vierkantor replied:
I have added a non-dependent `coe_fn_coe_base'`, which works here.

Vierkantor committed:
Fix `topology/fiber_bundle.lean`

Vierkantor committed:
Fix `topology/continouous_function/basic.lean`

Vierkantor committed:
Fix `topology/discrete_quotient.lean`

Vierkantor committed:
Fix `linear_algebra/basic.lean`

Vierkantor committed:
Fix `linear_algebra/basis.lean`

Vierkantor committed:
Fix `algebra/quaternion.lean`

Vierkantor committed:
Fix `algebra/non_unital_alg_hom.lean`

eric-wieser commented on src/algebra/quaternion.lean:
```diff
@@ -133,9 +133,9 @@ by refine_struct
-    npow := @npow_rec _ ⟨1⟩ ⟨(*)⟩ };
+    nsmul := @nsmul_rec _ ⟨(0 : ℍ[R, c₁, c₂])⟩ ⟨(+)⟩,
+    gsmul := @gsmul_rec _ ⟨(0 : ℍ[R, c₁, c₂])⟩ ⟨(+)⟩ ⟨has_neg.neg⟩,
+    npow := @npow_rec _ ⟨(1 : ℍ[R, c₁, c₂])⟩ ⟨(*)⟩ };
```
I don't get why this is necessary; surely there's no coe_sort or coe_fn involved here?

Vierkantor replied:
This is an elaboration order issue I think: the `0` gets elaborated too early, so its type is defaulted to `ℕ`, which later causes a type error because `nsmul_rec` should have type `ℕ → ℍ[R, c₁, c₂] → ℍ[R, c₁, c₂]` instead of `ℕ → ℕ → ℕ`.

eric-wieser replied:
I don't understand why the elaboration behavior changed though on expressions with no coercions

urkud replied:
There were some elaboration issues with new coercions. They were fixed by @gebner. It seems that these fixes had side effects.

eric-wieser replied:
It would be good to confirm these side effects are an expected compromise rather than an accident.

kim-em committed:
fix Module/adjunctions

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into coe-fn-backport

Vierkantor committed:
Fix `data/equiv/module.lean`

Vierkantor committed:
Fix `ring_theory/localization.lean`

Vierkantor committed:
Fix `topology/algebra/weak_dual_topology`

Vierkantor committed:
Fix `topology/homotopy.lean` (add param to `has_coe_to_fun`)

Vierkantor committed:
Fix `algebra/cateogry/FinVect.lean` (add param to `has_coe_to_sort`, `has_coe_to_fun`)

Vierkantor committed:
Add non-dependent versions of `coe_fn_coe_trans` and `coe_fn_coe_base`

This should help `rw` figure out the parameters, since it's a less complicated unification problem.

Vierkantor commented on src/logic/basic.lean:
```diff
@@ -82,16 +82,26 @@ lemma subtype.subsingleton (α : Sort*) [subsingleton α] (p : α → Prop) : su
+  (x : α) : @coe_fn α _ _ x = @coe_fn β _ _ x := rfl
+
+/-- Non-dependent version of `coe_fn_coe_base`, helps `rw` figure out the argument. -/
+theorem coe_fn_coe_base'
```
I would like to tag this as `@[simp]`, but that causes a timeout in `src/order/rel_iso.lean:465` which I don't quite know how to diagnose.

Vierkantor committed:
Use `coe_fn_coe_base'` to clean up proofs

Vierkantor committed:
Fix `linear_algebra/alternating.lean`: `coe_inj` has become `coe_injective`

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into coe-fn-backport

Vierkantor committed:
Fix `topology/continuous_function/algebra.lean` (`coe_fn_coe_base'`)

Vierkantor committed:
Fix `algebraic_topology/topological_simplex.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `data/complex/basic.lean` (number defaulting to ℕ)

Vierkantor committed:
Fix `analysis/normed_space/affine_isometry.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `topology/category/CompHaus/default.lean` (new `has_coe_to_sort`)

Vierkantor committed:
Fix `linear_algebra/special_linear_group.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `topology/category/Profinite/default.lean` (new `has_coe_to_sort`)

Vierkantor committed:
Fix `topology/partition_of_unity.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `analysis/normed_space/SemiNormedGroup.lean` (`has_coe_to_sort`, `coe_fn_coe_base'`)

Vierkantor committed:
Fix `number_theory/padics/padic_integers.lean` (number defaulting)

Vierkantor committed:
Fix `measure_theory/measure/measure_space_def.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `measure_theory/measure/measure_space.lean` (elaborator regression)

Vierkantor committed:
Fix `topology/continuous_function/stone_weierstrass.lean` (`coe_fn_coe_base'`)

Vierkantor committed:
Fix `ring_theory/fractional_ideal.lean` (number defaulting)

Vierkantor committed:
Fix `ring_theory/hahn_series.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `measure_theory/measure/finite_measure_weak_convergence.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `measure_theory/measure/vector_measure.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `measure_theory/measure/stieltjes.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `measure_theory/measure/content.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `geometry/manifold/partition_of_unity.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `geometry/manifold/derivation_bundle.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `geometry/manifold/algebra/left_invariant_derivation.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `measure_theory/function/conditional_expectation.lean` (`coe_fn_coe_base'`)

Vierkantor committed:
Fix `topology/metric_space/gromov_hausdorff_realized.lean` (proof got golfed!)

Vierkantor committed:
Fix some tests (new `has_coe_to_fun`)

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into coe-fn-backport

Vierkantor committed:
Fix timeout by splitting proof

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into coe-fn-backport

Vierkantor committed:
Fix `order/jordan_holder.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `linear_algebra/general_linear_group.lean` (new `has_coe_to_fun`, `coe_fn_coe_base'`)

Vierkantor committed:
Fix `topology/homotopy/basic.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `topology/homotopy/path.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Fix `ring_theory/algebraic_independent` (timeout, split proof)

Vierkantor committed:
Fix `ring_theory/perfection.lean` (timeout, by golfing and helping the elaborator)

Vierkantor committed:
Fix `tactic/norm_{fin,swap}.lean` (new `has_coe_to_fun`)

Vierkantor commented on src/tactic/norm_swap.lean:
```diff
@@ -31,7 +31,7 @@ example : equiv.swap 1 2 1 = 2 := by norm_num
 -/
 @[norm_num] meta def eval : expr → tactic (expr × expr) := λ e, do
-  (swapt, coe_fn_inst, fexpr, c) ← e.match_app_coe_fn <|> fail "did not get an app coe_fn expr",
+  (swapt, fun_ty, coe_fn_inst, fexpr, c) ← e.match_app_coe_fn <|> fail "did not get an app coe_fn expr",
```
To apply after the build succeeds (or fails...):
```suggestion
  (swapt, fun_ty, coe_fn_inst, fexpr, c) ← e.match_app_coe_fn
    <|> fail "did not get an app coe_fn expr",
```

Vierkantor commented on src/ring_theory/algebraic_independent.lean:
```diff
@@ -399,6 +398,39 @@ def algebraic_independent.mv_polynomial_option_equiv_polynomial_adjoin
+  hx.mv_polynomial_option_equiv_polynomial_adjoin (X (some i)) =
+    polynomial.C (hx.aeval_equiv (X i)) :=
+by rw [algebraic_independent.mv_polynomial_option_equiv_polynomial_adjoin_apply, aeval_X,
+      option.elim, polynomial.map_C, coe_coe, alg_hom.coe_to_ring_hom, alg_equiv.coe_alg_hom]
```
I expect all of these can be `@[simp]` lemmas. @ChrisHughes24 what do you think?

ChrisHughes24 replied:
I agree

Vierkantor committed:
Fix overlong line

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into coe-fn-backport

Vierkantor committed:
Fix `counterexamples/philips.lean` (new `has_coe_to_fun`)

Vierkantor committed:
Merge remote-tracking branch 'mathlib/master' into coe-fn-backport

Vierkantor committed:
Fix `analysis/box_integral/partition/additive.lean` (new `has_coe_to_fun`)
